dotnet/src/webdriver/Firefox/Preferences.cs

```diff
@@ -153,39 +160,46 @@ private static bool IsWrappedAsString(string value)
         }
 
-        private void SetPreferenceValue(string key, object value)
+        private void SetPreferenceValue(string key, JsonNode? value)
         {
             if (!this.IsSettablePreference(key))
             {
-                string message = string.Format(CultureInfo.InvariantCulture, "Preference {0} may not be overridden: frozen value={1}, requested value={2}", key, this.immutablePreferences[key], value.ToString());
+                string message = string.Format(CultureInfo.InvariantCulture, "Preference {0} may not be overridden: frozen value={1}, requested value={2}", key, this.preferences[key], value?.ToString());
                 throw new ArgumentException(message);
             }
 
-            string stringValue = value as string;
-            if (stringValue != null)
+            if (value is JsonValue jValue)
             {
-                if (IsWrappedAsString(stringValue))
+                if (jValue.TryGetValue(out string? stringValue))
                 {
-                    throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Preference values must be plain strings: {0}: {1}", key, value));
-                }
+                    if (IsWrappedAsString(stringValue))
+                    {
+                        throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, "Preference values must be plain strings: {0}: {1}", key, value));
+                    }
 
-                this.preferences[key] = string.Format(CultureInfo.InvariantCulture, "\"{0}\"", value);
-                return;
-            }
+                    this.preferences[key] = string.Format(CultureInfo.InvariantCulture, "\"{0}\"", value);
+                    return;
+                }
 
-            if (value is bool)
-            {
-                this.preferences[key] = Convert.ToBoolean(value, CultureInfo.InvariantCulture).ToString().ToLowerInvariant();
-                return;
-            }
+                if (jValue.TryGetValue(out bool boolValue))
+                {
+                    this.preferences[key] = boolValue ? "true" : "false";
+                    return;
+                }
 
-            if (value is int || value is long)
-            {
-                this.preferences[key] = Convert.ToInt32(value, CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture);
-                return;
+                if (jValue.TryGetValue(out int intValue))
+                {
+                    this.preferences[key] = intValue.ToString(CultureInfo.InvariantCulture);
+                    return;
+                }
+                if (jValue.TryGetValue(out long longValue))
+                {
+                    this.preferences[key] = longValue.ToString(CultureInfo.InvariantCulture);
+                    return;
+                }
             }
 
             throw new WebDriverException("Value must be string, int or boolean");
```

Comment:
I was trying to do research on the `webdriver_prefs.json` file format expectations, and I didn't see any decimal values anywhere. The current behavior will not only throw on doubles, but also `long`s. This PR already fixes `long` behavior (it will need to be `long.MaxValue + 1` before we throw), I can do the same for `double`'es.

The last primitive is `null`. That is still not allowed, right?

Reply:
Support for `double` has been added

Reply:
We should understand the purpose of this class. The good news is it is `internal` (just left a note for reviewers).

Reply:
I imagine this restriction is based on the possible values for Firefox profile preferences. Maybe Firefox will throw an exception if we try to set a wrong type of value?

We can accept any value here and let Firefox throw any values.

Should we accept any JSON values like objects or arrays, or should we allow only `true`, `false`, numbers, and strings? What about `null`?

Reply:
See, https://github.com/SeleniumHQ/selenium/blob/03386772307dde5dd063ef1bde35d8013132df94/dotnet/src/webdriver/Firefox/FirefoxProfile.cs#L137 (the same for int and string) - this entry point for public users. We are still manage "allowed" types.

Reply:
The entry point for the JSON parsing is the `webdriver_pref.json` file, which users can set values on. If their file has something like this:

```json
{
  "frozen": {
    "my-pref": [1, 2, 3]
  }
}
```

We will have to handle that somehow. Today, it throws the webdriver error "must be string, int, or boolean".

Reply:
Right, `webdriver_pref.json` is our file, fully manageable. We can parse it as is without any restrictions.

Reply:
After doing some research, I found out this:

> A value may either be integer (0, 2 etc.), boolean (true or false) or string (text, such as a file path)

https://kb.mozillazine.org/User.js_file#Adding_user.js_entries

For this reason, we can assume that the `webdriver_pref.json` will only ever be those values.

Reply:
OK! Final concern: we don't know validation rules on client side, and it is better to delegate validation to server side. I just want to say "deserialize as is because of it is json". Rules... should we replicate "unknown" rules on client side? I don't think so.

Reply:
Maintenance... it is effort.

Reply:
I agree. I always like "defense in depth", if someone has a bug (firefox, Selenium's server, misconfiguration of whoever wrote the `webdriver_pref.json` file...) I want this system to tell users clearly what went wrong.

But that cannot come at the expense of maintenance. I think the current state of this PR is simple enough, especially with the migration to `JsonDocument` in the latest commit.